gcc/config/aarch64/aarch64.h

```diff
@@ -407,11 +407,18 @@ enum class aarch64_feature : unsigned char {
    register.  GCC internally uses the poly_int variable aarch64_sve_vg
    instead.  */
 
+/* X18 reserved on windows for the TEB */
+#ifdef TARGET_CYGWIN64
```

Comment:
I would use TARGET_64BIT_MS_ABI
after that 
Option 1, define TARGET_64BIT_MS_ABI as 0 if it is not defined 
and use it in FIXED_REGISTERS
https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/510e10373ce9c8cbb78aeda293033a8f7465ab8c/gcc/config/nios2/nios2.h#L157

Option 2, rename X18 to FIXED_X18 and define it based on TARGET_64BIT_MS_ABI
https://github.com/Windows-on-ARM-Experiments/gcc-woarm64/blob/510e10373ce9c8cbb78aeda293033a8f7465ab8c/gcc/config/rs6000/rs6000.h#L815

Reply:
aarch64.h is included in many places at a very low level. This means that TARGET_64BIT_MS_ABI is often not present. TARGET_CYGWIN64 is currently defined in config.gcc making it always present. Obviously we could define TARGET_64BIT_MS_ABI in config.gcc as well. Is that what you are sugesting?

Reply:
In option 1, we can define TARGET_64BIT_MS_ABI as 0 if it is not defined in aarch64.h
In option 2, if TARGET_64BIT_MS_ABI is not defined FIXED_R13 will be 0

Reply:
Both those options do not give a correct result? 
The TARGET_64BIT_MS_ABI define is not availiable in aarch64.h reliably because in most cases aarch64-coff.h (where TARGET_64BIT_MS_ABI is currently defined) is not included first.

Reply:
I expect aarch64-coff.h will be included first for aarch64-mingw

Option 1: if TARGET_64BIT_MS_ABI is not defined, it will be defined as 0 in config/aarch64/aarch64.h and used in  FIXED_REGISTERS

Option 2: if TARGET_64BIT_MS_ABI is defined than FIXED_X18 is 1, otherwise FIXED_X18 is 0. and FIXED_X18  will be used in FIXED_REGISTERS

Reply:
it looks like you are referring to 
`#include "aarch64.h" 
`
in aarch64-coff.h
it might be it should not be included there

Reply:
What about adding (overriding) the entire `FIXED_REGISTERS`in `aarch64-coff.h`?

Reply:
Maybe I am not being clear :) 

When compiling our target, in some c files, aarch64.h is included and aarch64-coff.h is not. This means we cannot use any defines from aarch64-coff.h reliably in aarch64.h.

config.gcc is always included. That is why the TARGET_CYGWIN64 works. TARGET_CYGWIN64 is defined in config.gcc.

We could define TARGET_64BIT_MS_ABI in config.gcc. I asked above if people were recomending that. Seems like it might be best that way.

Reply:
I see. Then I'd suggest creating a file \`t-aarch64-coff\` (or similarly) that would include \`aarch64-coff.h\` that would be included in the resulting target config.